pkg/reconciler/testing/reconcile.go

```diff
@@ -231,6 +232,24 @@ func assertPopulatedSource(t *testing.T, src v1alpha1.AWSEventSource) {
 	assert.NotEmpty(t, src.GetStatus().Status.Conditions, "Provided source should have initialized conditions")
 }
 
+func nameKindAndResource(object runtime.Object) (string /*name*/, string /*kind*/, string /*resource*/) {
+	metaObj, _ := meta.Accessor(object)
+	name := metaObj.GetName()
+
+	var kind, resource string
+
+	switch object.(type) {
+	case *appsv1.Deployment:
+		kind = "Deployment"
+		resource = "deployments"
+	case *servingv1.Service:
+		kind = "Service"
+		resource = "services"
+	}
```

Comment:
I'm open to suggestions to do that in a more generic manner.

Reply:
I don't think there is an universal solution.
```go
metaObj.GetObjectKind()
```
could be use for kinds, but then you would need the _downcase-and-pluralize_ for the resource. Imho think it looks simpler as you wrote it.

Reply:
Even this doesn't work because API responses don't set TypeMeta, all you get is a typed object and you must use that to determine the API version and kind.

I always found that design weird, and there is a forever old k8s issue about it.